### PR TITLE
[System.Net.Http] Fix CFNetworkHandler when no data is sent

### PR DESCRIPTION
**Pull request by spouliot**
The handler receive an `Open` and the a `Close` event but never a
`DataAvailable` event.

So it _hangs_ if only a `200` status is received (since the headers
are not considered _data_) received from the web server.

This is because some initialization only occurs inside
`HandleHasBytesAvailableEvent` which is now called, on `Close`,
to ensure the initialization is done before completing the request.

This, along with a cancellation check, makes the test case from
https://github.com/xamarin/xamarin-macios/issues/5190 work properly
with this handler.

Fixes https://github.com/xamarin/xamarin-macios/issues/5289

### CODE REVIEW

**Comment by spouliot** [comment]
c.c. @mandel-macaque

**Comment by mandel-macaque** [comment]
For me, it looks good and the explanation makes perfect sense, if we got a header only response this can occur +1 although I can't give you a proper one in the mono repo :(

File: mcs/class/System.Net.Http/CFNetworkHandler.cs [hunk 1]
```diff
@@ -302,17 +302,19 @@ void HandleErrorEvent (object sender, CFStream.StreamEventArgs e)
...
 
 		void CloseStream (CFHTTPStream stream)
 		{
-			StreamBucket bucket;
-			if (streamBuckets.TryGetValue (stream.Handle, out bucket)) {
-				bucket.Close ();
-				streamBuckets.Remove (stream.Handle);
+			lock (streamBuckets) {
+				if (streamBuckets.TryGetValue (stream.Handle, out var bucket)) {
+					bucket.Close ();
```
Comment (marek-safar) [inline mcs/class/System.Net.Http/CFNetworkHandler.cs]: It might be better to do the Close outside of lock

**Comment by kochev** [comment]
Good day! We encountered the same problem. In test conditions, the problem is not reproduced. In production, we see about 3-5 crashes per day.
 Stack trace:
`CFNetworkHandler.HandleHasBytesAvailableEvent (System.Object sender, CoreFoundation.CFStream+StreamEventArgs e)
CFStream.OnHasBytesAvailableEvent (CoreFoundation.CFStream+StreamEventArgs args)
CFStream.OnCallback (CoreFoundation.CFStreamEventType type)
CFStream.OnCallback (System.IntPtr s, System.nint type, System.IntPtr info)
(wrapper native-to-managed) CoreFoundation.CFStream.OnCallback(intptr,System.nint,intptr)
(wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate)
UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName)
Application.Main (System.String[] args)
(wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate)
UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName)
Application.Main (System.String[] args)`

Is there any WA?
When will the fix go to the master branch?

**Pull request merged**